FR(Bedrock): Add unified token usage tracking for Bedrock LLMs

mohammadsubhani opened the pull request:
# PR Summary: Comprehensive Token Usage Tracking for Bedrock Autologging

## Related Issues / PRs
- #16233

## What Changes Are Proposed?
This PR adds **comprehensive token usage tracking** to MLflow's Bedrock autologging integration with **robust error handling**, **optimized parsing logic**, and **complete documentation**. It provides unified token usage extraction across all major Bedrock providers and APIs with extensive edge case coverage.

### Key Features
- **Multi-provider support:** Anthropic (Claude), AI21 (Jamba), Amazon Titan/Nova, Meta Llama
- **Full API coverage:** `invoke_model`, `invoke_model_with_response_stream`, `converse`, `converse_stream`
- **Unified format:** Standardized `{input_tokens, output_tokens, total_tokens}`
- **Streaming support:** Token usage captured and accumulated across streaming chunks
- **Error resilience:** Comprehensive handling of missing, malformed, or invalid usage data
- **Complete documentation:** User-facing docs with examples and API reference

### Technical Implementation
- **Optimized utilities:** `parse_complete_token_usage_from_response()` and `parse_partial_token_usage_from_response()`
- **Shared validation:** `_validate_usage_input()` and `_extract_token_value_by_keys()`
- **Streaming accumulation:** Buffered token usage across chunks with centralized parsing
- **Provider key mapping:** Constants supporting various key formats (`input_tokens`, `inputTokens`, `prompt_tokens`, etc.)
- **Span attributes:** Usage exposed via `SpanAttributeKey.CHAT_USAGE`
- **Enhanced error handling:** Try-catch blocks with debug logging for production resilience

### Code Quality Improvements
- **Consistent variable naming** (`usage_data` throughout, walrus operator consistency)
- **Strong type hints** (`Optional[dict[str, int]]`, `Sequence[str]`, proper validation)
- **Reduced duplication** through shared utilities and consolidated edge case data
- **Early returns** for performance optimization
- **Comprehensive test coverage** with 32+ edge cases

## How Is This PR Tested?

### Automated Tests
- **All existing tests preserved** - no functionality removed
- **32+ new edge case tests** covering comprehensive scenarios:
  - **Token usage edge cases:** Missing usage, empty dicts, malformed data, None values, string values, very large values, zero values
  - **Streaming edge cases:** Usage in different chunks, chunk updates, malformed streaming data
  - **Provider variations:** Different response formats across Anthropic, AI21, Amazon, Meta
  - **API coverage:** All 4 Bedrock APIs with token usage validation
- **Enhanced existing tests** with token usage validation using parameterized expected values
- **Consolidated test data** to reduce duplication while maintaining comprehensive coverage

### Manual Testing
- Verified token extraction across all supported providers
- Confirmed streaming token accumulation behavior
- Validated documentation examples with real trace data
- Tested error handling with malformed responses

## Documentation Update
**Yes** — Updated:
- **Complete user documentation** in `docs/docs/genai/tracing/integrations/listing/bedrock.mdx`:
  - New "Token Usage" section with detailed explanation
  - Code examples showing trace inspection and token access
  - Supported models, APIs, and provider coverage
  - Clear examples of expected output format
- **Enhanced code documentation:** Comprehensive docstrings and inline comments
- **Function documentation:** Clear parameter descriptions and return value specifications

## Release Notes (User-Facing Description)
Added comprehensive token usage tracking to MLflow's Bedrock autologging integration. Users can now monitor token consumption across all major Bedrock providers (Anthropic, AI21, Amazon, Meta) and APIs (invoke_model, converse, streaming). Token usage appears in the `mlflow.chat.tokenUsage` span attribute and trace-level `token_usage` field. Includes robust error handling for production use and complete documentation with examples.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [x] `integrations/sagemaker`: SageMaker integrations (Bedrock is AWS service)

#### How should the PR be classified in the release notes? Choose one:
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?
- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

**Reasoning:** This is a new feature that adds token usage tracking functionality without breaking existing behavior. It enhances the Bedrock integration with new capabilities that users can discover and adopt gradually.

---

## Files Changed

| File | Summary |
|------|---------|
| `mlflow/bedrock/_autolog.py` | Enhanced with token usage parsing, consistent error handling, and walrus operator usage |
| `mlflow/bedrock/stream.py` | Streaming token accumulation with buffering, improved error handling, and consistent naming |
| `mlflow/bedrock/utils.py` | Core parsing utilities with shared validation, provider key mapping, and comprehensive error handling |
| `docs/docs/genai/tracing/integrations/listing/bedrock.mdx` | Complete "Token Usage" documentation with examples, API coverage, and usage instructions |
| `tests/bedrock/test_bedrock_autolog.py` | Enhanced existing tests + 32+ new edge case tests with consolidated test data and comprehensive coverage |

## Code Quality Metrics
- **New functionality:** Token usage tracking for all Bedrock APIs and providers
- **New tests:** 32+ comprehensive edge case tests
- **Test coverage:** All APIs, providers, edge cases, and error conditions
- **Code improvements:** Optimized parsing, consistent naming, reduced duplication
- **Error handling:** Production-ready with comprehensive exception handling
- **No breaking changes:** All existing functionality preserved and enhanced

B-Step62 commented on mlflow/bedrock/_autolog.py:
```diff
@@ -129,6 +139,71 @@ def _parse_invoke_model_response_body(response_body: StreamingBody) -> Union[dic
+        # Handle Anthropic format
+        if "usage" in response_body and isinstance(response_body["usage"], dict):
+            usage = response_body["usage"]
+            if "input_tokens" in usage and "output_tokens" in usage:
+                return {
+                    TokenUsageKey.INPUT_TOKENS: usage["input_tokens"],
+                    TokenUsageKey.OUTPUT_TOKENS: usage["output_tokens"],
+                    TokenUsageKey.TOTAL_TOKENS: usage.get(
+                        "total_tokens",
+                        usage["input_tokens"] + usage["output_tokens"],
+                    ),
+                }
+            # Handle AI21 format
+            elif "prompt_tokens" in usage and "completion_tokens" in usage:
+                return {
+                    TokenUsageKey.INPUT_TOKENS: usage["prompt_tokens"],
+                    TokenUsageKey.OUTPUT_TOKENS: usage["completion_tokens"],
+                    TokenUsageKey.TOTAL_TOKENS: usage.get(
+                        "total_tokens",
+                        usage["prompt_tokens"] + usage["completion_tokens"],
+                    ),
+                }
+            # Handle Amazon Nova format
+            elif "inputTokens" in usage and "outputTokens" in usage:
+                return {
+                    TokenUsageKey.INPUT_TOKENS: usage["inputTokens"],
+                    TokenUsageKey.OUTPUT_TOKENS: usage["outputTokens"],
+                    TokenUsageKey.TOTAL_TOKENS: usage.get(
+                        "totalTokens",
+                        usage["inputTokens"] + usage["outputTokens"],
+                    ),
```
Can we simplify the logic like

```
if "input_tokens" in usage and "output_tokens" in usage:
    input_token_key, output_token_key = "input_tokens", "output_tokens"
elif "prompt_tokens" in usage and "completion_tokens" in usage:
        input_token_key, output_token_key = "prompt_tokens", "completion_tokens"

return {
    TokenUsageKey.INPUT_TOKENS: usage[input_token_key]
   ...
}
```

```

B-Step62 commented on mlflow/bedrock/stream.py:
```diff
@@ -65,12 +66,71 @@ def _end_span(self):
+    def _buffer_usage(self, chunk: dict):
+        """Buffer token usage information from streaming chunks."""
+        try:
+            # Handle usage in message_start event (Anthropic format)
```
Do we need to handle non-Anthoropic format here?

B-Step62 commented on mlflow/bedrock/stream.py:
```diff
@@ -65,12 +66,71 @@ def _end_span(self):
+        """Buffer token usage information from streaming chunks."""
+        try:
+            # Handle usage in message_start event (Anthropic format)
+            if chunk.get("type") == "message_start" and "usage" in chunk.get("message", {}):
+                usage = chunk["message"]["usage"]
+                if "input_tokens" in usage:
+                    self._usage_buffer[TokenUsageKey.INPUT_TOKENS] = usage["input_tokens"]
+                if "output_tokens" in usage:
+                    self._usage_buffer[TokenUsageKey.OUTPUT_TOKENS] = usage["output_tokens"]
+                    _logger.debug(
+                        f"[InvokeModelStreamWrapper] Set output_tokens to "
+                        f"{usage['output_tokens']} from message_start"
+                    )
+
+            # Handle usage in message_delta event (Anthropic format)
+            elif chunk.get("type") == "message_delta" and "usage" in chunk.get("delta", {}):
+                usage = chunk["delta"]["usage"]
+                # Always update output_tokens if present (take the latest value)
+                if "output_tokens" in usage:
+                    self._usage_buffer[TokenUsageKey.OUTPUT_TOKENS] = usage["output_tokens"]
+                    _logger.debug(
+                        f"[InvokeModelStreamWrapper] Updated output_tokens to "
+                        f"{usage['output_tokens']} from message_delta"
+                    )
```
Can we consolidate duplicated logic here? The output token extraction logic is same.

B-Step62 commented on mlflow/bedrock/stream.py:
```diff
@@ -65,12 +66,71 @@ def _end_span(self):
+                output_tokens = self._usage_buffer.get(TokenUsageKey.OUTPUT_TOKENS, 0)
+                self._usage_buffer[TokenUsageKey.TOTAL_TOKENS] = input_tokens + output_tokens
+            usage_dict = {
+                TokenUsageKey.INPUT_TOKENS: self._usage_buffer.get(TokenUsageKey.INPUT_TOKENS),
+                TokenUsageKey.OUTPUT_TOKENS: self._usage_buffer.get(TokenUsageKey.OUTPUT_TOKENS),
+                TokenUsageKey.TOTAL_TOKENS: self._usage_buffer.get(
+                    TokenUsageKey.TOTAL_TOKENS,
+                    (
+                        self._usage_buffer.get(TokenUsageKey.INPUT_TOKENS, 0)
+                        + self._usage_buffer.get(TokenUsageKey.OUTPUT_TOKENS, 0)
+                    ),
+                ),
```
Why do we need to create another dictionary here? Isn't `_usage_buffer` already compatible form of usage attribute value?

B-Step62 commented on mlflow/bedrock/stream.py:
```diff
@@ -106,6 +166,23 @@ def _close(self):
+        # Record token usage if provided in the response
+        usage = converse_response.get("usage")
+        if isinstance(usage, dict):
+            usage_dict = {
+                TokenUsageKey.INPUT_TOKENS: usage.get("inputTokens"),
+                TokenUsageKey.OUTPUT_TOKENS: usage.get("outputTokens"),
+                TokenUsageKey.TOTAL_TOKENS: usage.get(
+                    "totalTokens", (usage.get("inputTokens", 0) + usage.get("outputTokens", 0))
+                ),
+            }
+            # Only set if we have valid token counts
+            if (
+                usage_dict[TokenUsageKey.INPUT_TOKENS] is not None
+                and usage_dict[TokenUsageKey.OUTPUT_TOKENS] is not None
+            ):
+                self._span.set_attribute(SpanAttributeKey.CHAT_USAGE, usage_dict)
```
Let's abstract this repeated logic to utility function.

B-Step62 commented on tests/bedrock/test_bedrock_autolog.py:
```diff
@@ -18,6 +18,16 @@
 _IS_CONVERSE_API_AVAILABLE = Version(boto3.__version__) >= Version("1.35")
 
 
+@pytest.fixture(autouse=True)
```
Will the test fail if we don't do this? I guess not because we have reset fixture in `tests/conftest.py`?

B-Step62 commented on tests/bedrock/test_bedrock_autolog.py:
```diff
@@ -220,6 +235,122 @@ def test_bedrock_autolog_invoke_model_llm(model_id, llm_request, llm_response):
+            )
+
+
+def test_bedrock_autolog_token_usage_anthropic():
```
These test cases overlap with the one above.

B-Step62 commented on tests/bedrock/test_bedrock_autolog.py:
```diff
@@ -220,6 +235,122 @@ def test_bedrock_autolog_invoke_model_llm(model_id, llm_request, llm_response):
+    # Check token usage
+    assert SpanAttributeKey.CHAT_USAGE in span.attributes
+    usage = span.attributes[SpanAttributeKey.CHAT_USAGE]
+    if "usage" in llm_response:
```
Let's simplify this logic as suggested in the other comment.

B-Step62 commented on mlflow/bedrock/stream.py:
```diff
@@ -65,12 +66,90 @@ def _end_span(self):
+            # For streaming, we need to handle partial updates where only some token counts
+            # are provided. Extract individual token counts and update buffer directly.
+
+            def _pick(d: dict, *names):
+                for n in names:
+                    if n in d:
+                        return d[n]
+                return None
```
```suggestion
            def _pick(d: dict[str, Any], names: list[str]) -> int:
                return next((d[name] for name in names if name in d), None)
```

Let's not use variable-length argument if it's not necessary.

B-Step62 commented on mlflow/bedrock/stream.py:
```diff
@@ -65,12 +66,90 @@ def _end_span(self):
+                "totalTokens",
+            )
+
+            # Update buffer with any non-None values (partial updates are allowed)
```
Q: Is there any case where the same key (e.g. "input_tokens") show up in multiple chunks?

B-Step62 commented on mlflow/bedrock/stream.py:
```diff
@@ -65,12 +66,90 @@ def _end_span(self):
+            )
+
+        try:
+            if chunk.get("type") == "message_start" and "usage" in chunk.get("message", {}):
```
Do we need to check chunk type here? Shall we extract token usage whenever it presents in chunk?

B-Step62 commented on mlflow/bedrock/utils.py:
```diff
@@ -41,3 +42,107 @@ def wrapper(original, self, *args, **kwargs):
+def build_token_usage_dict(
+    *,
+    raw_usage: Optional[dict[str, int]] = None,
+    input_tokens: Optional[int] = None,
+    output_tokens: Optional[int] = None,
+    total_tokens: Optional[int] = None,
```
These parameters are not used in anywhere.

B-Step62 replied:
@mohammadsubhani  We should have two separate functions then. Using the same function with exclusive two set of parameters is anti pattern.

B-Step62 commented on mlflow/bedrock/utils.py:
```diff
@@ -41,3 +42,107 @@ def wrapper(original, self, *args, **kwargs):
+
+
+def build_token_usage_dict(
+    *,
```
Why do we need to prevent positional argument here?

mohammadsubhani replied:
**The `*` (keyword-only arguments) prevents positional arguments for clarity and safety:**

**Why it's needed:**
- All parameters are `Optional[int]`, so it's easy to mix up their order
- The function supports **two distinct usage patterns** that would be ambiguous with positional args:
  1. **Raw usage dict parsing** - `build_token_usage_dict(raw_usage=response.get("usage"))`
  2. **Individual token counts** - `build_token_usage_dict(input_tokens=10, output_tokens=20)`

**Real usage examples from the codebase:**

**Pattern 1 - Raw usage dict:**
```python
build_token_usage_dict(raw_usage=response_body.get("usage", response_body))
```

**Pattern 2 - Individual token counts:**
```python
build_token_usage_dict(
    input_tokens=self._usage_buffer.get(TokenUsageKey.INPUT_TOKENS),
    output_tokens=self._usage_buffer.get(TokenUsageKey.OUTPUT_TOKENS),
    total_tokens=self._usage_buffer.get(TokenUsageKey.TOTAL_TOKENS),
)
```

**Without keyword-only args, these would be ambiguous:**
```python
build_token_usage_dict(usage_dict, 10, 20)  # raw_usage + input_tokens + output_tokens?
build_token_usage_dict(10, 20, 30)          # input_tokens + output_tokens + total_tokens?
```

The keyword-only requirement ensures intent is always clear and prevents parameter order confusion.

B-Step62 replied:
As I mentioned below, the problem is the function consolidate two different usage into one. Imo the better solution is to split into two functions with clearer responsibility, not enforcing kwargs.

B-Step62 commented on mlflow/bedrock/utils.py:
```diff
@@ -41,3 +42,107 @@ def wrapper(original, self, *args, **kwargs):
+    The returned dict uses :pydata:`mlflow.tracing.constant.TokenUsageKey` entries
+    (``input_tokens``, ``output_tokens``, ``total_tokens``).
+
+    Parameters
+    ----------
```
Please keep the docstring style consistent with the code base. We are using Google docstring style.

B-Step62 commented on mlflow/bedrock/utils.py:
```diff
@@ -41,3 +42,107 @@ def wrapper(original, self, *args, **kwargs):
+        can be determined, otherwise ``None``.
+    """
+
+    def _pick(d: dict, *names):
+        for n in names:
+            if n in d:
+                return d[n]
+        return None
+
+    if raw_usage:
```
Repetition of the logic in InvokeModelStreamWrapper. Can you extract this as a separate util?

B-Step62 commented on mlflow/bedrock/utils.py:
```diff
@@ -41,3 +42,107 @@ def wrapper(original, self, *args, **kwargs):
+    }
+
+
+def set_chat_usage_if_valid(span, usage_dict: Optional[dict]):
```
nit: I don't think we need a separate utility function just for checking None.

B-Step62 commented on tests/bedrock/test_bedrock_autolog.py:
```diff
@@ -220,6 +225,21 @@ def test_bedrock_autolog_invoke_model_llm(model_id, llm_request, llm_response):
+    assert SpanAttributeKey.CHAT_USAGE in span.attributes
+    usage = span.attributes[SpanAttributeKey.CHAT_USAGE]
+
+    from mlflow.bedrock.utils import build_token_usage_dict
```
No need to do lazy import. Can we move it up to the top-level?

mohammadsubhani commented on tests/bedrock/test_bedrock_autolog.py:
```diff
@@ -878,7 +924,7 @@ def _event_stream(raw_response, chunk_size=10):
     yield {"messageStop": {"stopReason": "end_turn"}}
 
-    yield {"metadata": {"usage": raw_response["usage"], "metrics": {"latencyMs": 551}}}
+    yield {"metadata": {"usage": raw_response.get("usage"), "metrics": {"latencyMs": 551}}}
```
nit: defensive programming improvement

mohammadsubhani commented:
Hi @B-Step62, I have made the necessary changes. Can you kindly review them? Thank you!

B-Step62 commented on mlflow/bedrock/utils.py:
```diff
@@ -1,12 +1,60 @@
+]
+
+
+def _pick(d: dict[str, Any], *names: str) -> Optional[int]:
```
Let's give more descriptive name as this is now a public method.

B-Step62 commented on mlflow/bedrock/stream.py:
```diff
@@ -65,12 +73,55 @@ def _end_span(self):
+        """Buffer token usage information from streaming chunks, always using the latest value."""
+
+        def _extract_and_buffer_tokens(usage: dict[str, Any]):
+            for k, v in {
```
```suggestion
    for key in TokenUsageKey.all_keys():
        v = _pick(usage, mlflow_key_to_bedrock_keys.get(key))
```

B-Step62 commented on mlflow/bedrock/stream.py:
```diff
@@ -65,12 +73,55 @@ def _end_span(self):
+            }.items():
+                if v is not None:
+                    self._usage_buffer[k] = v  # Always overwrite with the latest value
+            _logger.debug(
```
nit: Do we need this debug log?

B-Step62 commented on mlflow/bedrock/stream.py:
```diff
@@ -106,6 +157,14 @@ def _close(self):
 
+        # Build a standardized usage dict and set it on the span if valid
+        if (
+            usage_dict := build_token_usage_dict(raw_usage=converse_response.get("usage"))
```
nit: Mixing `:=` and ternary operator in if condition is a bit too complex (the order of evaluation is not obvious). Let's extract them into stand-alone lines.

mohammadsubhani commented:
Hi @B-Step62  , I have made the necessary changes and addressed your feedback. Can you kindly review it? Thank you!

B-Step62 commented on mlflow/bedrock/utils.py:
```diff
@@ -41,3 +69,93 @@ def wrapper(original, self, *args, **kwargs):
+        >>> parse_token_usage_from_response({"inputTokens": 10, "completionTokens": 15})
+        {'input_tokens': 10, 'output_tokens': 15, 'total_tokens': 25}
+    """
+    if require_full_usage:
```
@mohammadsubhani Let's split the function rather than adding a control flag `require_full_usage`. There is no common logic between two branch.

B-Step62 commented on mlflow/bedrock/stream.py:
```diff
@@ -62,20 +66,75 @@ def _end_span(self):
     def _close(self):
+        """Finalize the streaming span with accumulated token usage data."""
+        # Build a standardized usage dict from buffered data using the utility function
+        usage_dict = parse_token_usage_from_response(self._usage_buffer, require_full_usage=True)
+
+        if usage_dict:
+            self._span.set_attribute(SpanAttributeKey.CHAT_USAGE, usage_dict)
+
```
```suggestion
        if usage_dict := parse_token_usage_from_response(self._usage_buffer, require_full_usage=True)
            self._span.set_attribute(SpanAttributeKey.CHAT_USAGE, usage_dict)

```

nit

Subhanimuhammed committed:
Add comprehensive token usage tracking for AWS Bedrock

This commit implements comprehensive token usage tracking for all AWS Bedrock APIs
including invoke_model, invoke_model_with_response_stream, converse, and converse_stream.

Features:
- Token usage parsing for all major Bedrock providers (Anthropic, AI21, Amazon Nova/Titan, Meta Llama)
- Support for streaming token accumulation across chunks
- Comprehensive edge case handling (missing usage, malformed data, etc.)
- 32+ test cases covering all edge cases and provider variations
- Complete documentation with examples

Technical improvements:
- Optimized token extraction logic with early returns
- Shared validation functions to reduce code duplication
- Enhanced error handling with proper logging
- Consistent variable naming across all modules
- Type-safe implementations with accurate type hints

The implementation handles various provider response formats seamlessly and
provides robust error handling for production use.

Signed-off-by: mohammadsubhani <subhanimuhammed@gmail.com>

mohammadsubhani commented on mlflow/bedrock/stream.py:
```diff
@@ -60,15 +65,67 @@ def _end_span(self):
+            return parse_partial_token_usage_from_response(usage)
+        return None
+    except (KeyError, TypeError, AttributeError) as e:
+        _logger.debug(f"Failed to extract token usage from chunk: {e}")
```
Added logger to record why the extraction failed without breaking the stream.

mohammadsubhani commented on mlflow/bedrock/_autolog.py:
```diff
@@ -60,6 +60,31 @@ def patch_bedrock_runtime_client(client_class: type[BaseClient]):
+        return None
+    except (KeyError, TypeError, ValueError) as e:
+        _logger.debug(f"Failed to parse token usage from response: {e}")
+        return None
```
Added logger to record why the extraction failed without breaking the stream.

mohammadsubhani commented on mlflow/bedrock/utils.py:
```diff
@@ -1,12 +1,62 @@
+                - output_tokens / outputTokens: Output token count
+                - completion_tokens / completionTokens: Also mapped as output token count
+                - total_tokens / totalTokens: Total token count (input + output)"""
+
```
consolidated as a common usage_docs, which can be used to extend future additions as well.

B-Step62 reviewed (approved):
LGTM!

Subhanimuhammed committed:
Fix formatting in bedrock.mdx documentation

Signed-off-by: mohammadsubhani <subhanimuhammed@gmail.com>
